pkg/resource/deploy/plan_executor.go

```diff
@@ -0,0 +1,224 @@
+		atomic.SwapUint32(&pe.sawCancel, 1)
+		pe.stepExecCancel.Cancel()
+	}()
+
+	// The second one polls for incoming source events and writes them to the `incomingEvents` channel,
+	// so the man loop can `select` on it.
+	type nextEvent struct {
+		Event SourceEvent
+		Error error
+	}
+
+	incomingEvents := make(chan nextEvent)
+	go func() {
+		for {
+			event, sourceErr := pe.src.Next()
+			incomingEvents <- nextEvent{event, sourceErr}
+		}
+	}()
+
+	// The main loop. We'll continuously select for incoming events and the cancellation signal. There are
+	// a three ways we can exit this loop:
+	//  1. The SourceIterator sends us a `nil` event. This means that we're done processing source events and
+	//     we should begin processing deletes.
+	//  2. The SourceIterator sends us an error. This means some error occurred in the source program and we
+	//     should bail.
+	//  3. The stepExecCancel cancel context gets canceled. This means some error occurred in the step executor
+	//     and we need to bail. This can also happen if the user hits Ctrl-C.
+outer:
+	for {
+		logging.V(planExecutorLogLevel).Infof("PlanExecutor.Execute(...): waiting for incoming events")
+		select {
+		case event := <-incomingEvents:
+			logging.V(planExecutorLogLevel).Infof("PlanExecutor.Execute(...): incoming event")
+			if event.Error != nil {
+				logging.V(planExecutorLogLevel).Infof("PlanExecutor.Execute(...): saw incoming error: %v", event.Error)
+				pe.cancelDueToError()
+				break outer
+			}
+
+			if event.Event == nil {
+				logging.V(planExecutorLogLevel).Infof("PlanExecutor.Execute(...): saw nil event, beginning termination")
+
+				// TODO[pulumi/pulumi#1625] Today we lack the ability to parallelize deletions. We have all the
+				// information we need to do so (namely, a dependency graph). `GenerateDeletes` returns a single
+				// chain of every delete that needs to be executed.
+				deletes := pe.stepGen.GenerateDeletes()
+				pe.stepExec.Execute(deletes)
+
+				// Signal completion to the step executor. It'll exit once it's done retiring all of the steps in
+				// the chain that we just gave it.
+				pe.stepExec.SignalCompletion()
+				logging.V(planExecutorLogLevel).Infof("PlanExecutor.Execute(...): completed deletes, exiting loop")
+				break outer
+			}
+
+			pe.handleSingleEvent(event.Event)
+		case <-pe.stepExecCancel.Context().Canceled():
+			logging.V(planExecutorLogLevel).Infof("PlanExecutor.Execute(...): context canceled")
+			pe.cancelDueToError()
+			break outer
+		}
+	}
+
+	logging.V(planExecutorLogLevel).Infof("PlanExecutor.Execute(...): exited event loop, waiting for completion")
+	pe.stepExec.WaitForCompletion()
+	logging.V(planExecutorLogLevel).Infof("PlanExecutor.Execute(...): step executor has completed")
+
+	// To provide the best error message we can, we've kept track of whether or not we were successful and, if we
+	// were not, if we failed because of a cancel or because the step executor died.
+	if pe.canceled() {
```

Comment:
Should `errored` trump `canceled`?

Reply:
I picked "no" to that question because I figured that, if somebody is canceling a plan, they are making an explicit gesture to do so and it does not matter whether or not the plan has seen an error yet or not. The most accurate message we can give when somebody presses Ctrl+C is that they canceled the plan, since that's exactly what they did. It seemed strange to me that a user could get different messages when they Ctrl+C a plan depending on whether or not the plan was successful up until that point.